internal/docstore/mongodocstore/mongo.go

```
@@ -454,6 +488,166 @@ func (c *collection) docID(doc driver.Document) (interface{}, error) {
+		}
+	}
+	return alerr
+}
+
+func (c *collection) unorderedBulkWrite(ctx context.Context, iactions []indexedAction) driver.ActionListError {
+	var (
+		alerr              driver.ActionListError
+		models             []mongo.WriteModel
+		newIDs             = map[int]interface{}{} // new IDs for Create, keyed by position in iactions
+		nDeletes, nMatches int64
+	)
+	for i, ia := range iactions {
+		a := ia.action
+		var m mongo.WriteModel
+		var err error
+		var newID interface{}
+		switch a.Kind {
+		case driver.Create:
+			m, newID, err = c.newCreateModel(ia.action)
+			if newID != nil {
+				newIDs[i] = newID
+			}
+		case driver.Delete:
+			m, err = c.newDeleteModel(a)
+			if err != nil {
+				nDeletes++
+			}
+		case driver.Replace, driver.Put:
+			m, err = c.newReplaceModel(a, a.Kind == driver.Put)
+			if err != nil && a.Kind != driver.Put {
+				nMatches++
+			}
+		case driver.Update:
+			m, err = c.newUpdateModel(a)
+			if err != nil && m != nil {
+				nMatches++
+			}
+
+		default:
+			err = gcerr.Newf(gcerr.Internal, nil, "bad action %+v", a)
+		}
+		if err != nil {
+			alerr = append(alerr, indexedError{ia.index, err})
+		} else if m != nil { // m can be nil for a no-op update
+			models = append(models, m)
+		}
+	}
+	res, err := c.coll.BulkWrite(ctx, models, options.BulkWrite().SetOrdered(false))
+	if err != nil {
+		bwe, ok := err.(*mongo.BulkWriteException)
+		if !ok { // assume everything failed with this error
+			return append(alerr, indexedError{-1, err})
+		}
+		// The returned indexes of the WriteErrors are wrong. See https://jira.mongodb.org/browse/GODRIVER-1028.
+		// Until it's fixed, use negative values for the indexes in the errors we return.
+		for _, w := range bwe.WriteErrors {
+			alerr = append(alerr, indexedError{-1, gcerr.Newf(translateMongoCode(w.Code), nil, "%s", w.Message)})
+		}
+		return alerr
+	}
+	if res.DeletedCount != nDeletes {
+		alerr = append(alerr, indexedError{-1, gcerr.Newf(gcerr.NotFound, nil, "some delete failed")})
```

If user redo the same deletes when this happen, will Mongo succeed for the remaining ones and silently ignore those that are already deleted? If so we can suggest the user to retry it.

---

It's much more likely that the documents just don't exist. If there were a real error, like a network issue, then I'm pretty sure the call would return an error.